### PR TITLE
GH-677: Create Helm charts for Cromwell Auth Proxy for GOTC

### PR DESCRIPTION
**Pull request by rexwangcc**
https://broadinstitute.atlassian.net/browse/GH-677

Review instructions:
Currently for the sake of testing we point the proxy to pharma5 Cromwell, the URL is here: https://cromwell-gotc-auth.gotc-dev.broadinstitute.org/

Make sure:
- https://cromwell-gotc-auth.gotc-dev.broadinstitute.org/engine/v1/version returned the correct version.
- With `curl -X GET "https://cromwell-gotc-auth.gotc-dev.broadinstitute.org/api/workflows/v1/query?status=Aborted" -L -H "authorization:Bearer $(gcloud auth print-access-token)"`, you should either see the results or a Auth error that returned by **`cromwell-pharma5.gotc-prod.broadinstitute.org`**, depends on whether you have permission to pharma5 Cromwell proxy.
- For the actual values used by this Helm chart installation, ask @Fab-T or @hjfbynara

### CODE REVIEW

File: helm/cromwell_openidc/templates/configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    LogLevel ${LOG_LEVEL}
+
+    Header unset X-Frame-Options
+    Header always set X-Frame-Options SAMEORIGIN
+
+    ProxyTimeout ${PROXY_TIMEOUT}
+    OIDCOAuthTokenIntrospectionInterval 60
+
+    <VirtualHost _default_:${HTTPD_PORT}>
+        ServerAdmin ${SERVER_ADMIN}
```
Comment (Fab-T) [inline helm/cromwell_openidc/templates/configmap.yaml]: There is some duplicate declaration of variables , some already defined on top for all VH, that I don't think each VH should override ... for customLog for ex you should keep the one on top level already setup for proxy mode.

File: helm/cromwell_openidc/templates/configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        RewriteCond %{REQUEST_URI}  !^/(version|status) [NC]
+
+        ErrorLog /dev/stdout
+        CustomLog "/dev/stdout" combined
+
+        <Location ${PROXY_PATH}>
+            RewriteCond %{REQUEST_METHOD} OPTIONS
+            RewriteRule ^(.*)$ $1 [R=204,L]
+
+            <Limit OPTIONS>
```
Comment (Fab-T) [inline helm/cromwell_openidc/templates/configmap.yaml]: For security reasons you should limit the actions allowed to HEAD POST GET , usually enough...and I am guessing Cromwell accepts POST only ?!

File: helm/cromwell_openidc/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        app: {{ include "cromwell-openidc.fullname" . }}
+        component: {{ include "cromwell-openidc.fullname" . }}
+    spec:
+      serviceAccountName: {{ include "cromwell-openidc.serviceAccountName" . }}
+      containers:
+      - name: {{ .Chart.Name }}
+        image: "{{ .Values.image.repository }}:{{ .Values.image.version }}"
+        imagePullPolicy: {{ .Values.image.pullPolicy }}
+        ports:
+          - name: http
```
Comment (Fab-T) [inline helm/cromwell_openidc/templates/deployment.yaml]: Question : 
if the service is a proxy for auth, the protocol should be secured - HTTPS / 443 / TLS ?
Not sure :  may be this container end point is only available within the cluster and the secure access will be defined with ingress.

File: helm/cromwell_openidc/templates/service.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ include "cromwell-openidc.fullname" . }}
+  labels:
+    {{- include "cromwell-openidc.labels" . | nindent 4 }}
+spec:
+  type: {{ .Values.service.type }}
+  ports:
```
Comment (Fab-T) [inline helm/cromwell_openidc/templates/service.yaml]: Same concern here with service exposed over HTTP where it should be secured : HTTPS / TLS

File: helm/cromwell_openidc/templates/deployment.yaml (hunk 3 above)
Comment (rexwangcc) [reply helm/cromwell_openidc/templates/deployment.yaml]: Thanks, yeah I think it'd be safer to have HTTPS protection. I'm not sure if this is only available within the cluster either. Will need to double check with others.

**Comment by hjfbynara** [comment]
Oh one comment - what is the general guidelines on squashing commits before PR?   Opening a PR in order to seek feed back - I can see having all the commits there so you can see the progression.  But once you have a good version that you think is ready to merge - I find all the commits kind of messy.  So I generally do a squash before my final PR.  But I will go with what ever guidelines that Hornet team has put in place.

File: helm/auth_proxy/templates/configmap.yaml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        RewriteEngine On
+        RewriteCond  %{HTTP:X-Forwarded-Proto} !https
+        RewriteCond %{REQUEST_URI}  !^/engine/v1/(version|status) [NC]
+        RewriteRule (.*) https://${SERVER_NAME}%{REQUEST_URI}
+
+        RewriteCond %{REQUEST_METHOD} OPTIONS
+        RewriteRule ^(.*)$ $1 [R=204,L]
+
+
+        DocumentRoot /app
```
Comment (Fab-T) [inline helm/auth_proxy/templates/configmap.yaml]: I think we can remove everything related to /app line 42 - 50 as it's empty and reading the doc from /openidc-baseimage exists should you need to mount a volume for your app. And we don't have any app.

File: helm/auth_proxy/templates/configmap.yaml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            </Limit>
+
+            ${AUTH_TYPE}
+            ${AUTH_LDAP_URL}
+            ${AUTH_LDAP_GROUP_ATTR}
+            ${AUTH_LDAP_BIND_DN}
+            ${AUTH_LDAP_BIND_PASSWORD}
+            ${AUTH_REQUIRE}
+
+            <Limit OPTIONS>
```
Comment (Fab-T) [inline helm/auth_proxy/templates/configmap.yaml]: declared twice, should remove one

File: helm/auth_proxy/templates/configmap.yaml [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            </Limit>
+
+            ${AUTH_TYPE2}
+            ${AUTH_LDAP_URL2}
+            ${AUTH_LDAP_GROUP_ATTR2}
+            ${AUTH_LDAP_BIND_DN2}
+            ${AUTH_LDAP_BIND_PASSWORD2}
+            ${AUTH_REQUIRE2}
+
+            <Limit OPTIONS>
```
Comment (Fab-T) [inline helm/auth_proxy/templates/configmap.yaml]: same should remove one

File: helm/cromwell_openidc/templates/deployment.yaml (hunk 3 above)
Comment (Fab-T) [reply helm/cromwell_openidc/templates/deployment.yaml]: I realized TLS is handled by ingress - should be good !

**Comment by Fab-T** [comment]
Very nice job !
Tested and validated.

**Comment by rexwangcc** [comment]
> Oh one comment - what is the general guidelines on squashing commits before PR? Opening a PR in order to seek feed back - I can see having all the commits there so you can see the progression. But once you have a good version that you think is ready to merge - I find all the commits kind of messy. So I generally do a squash before my final PR. But I will go with what ever guidelines that Hornet team has put in place.

That's a good point to bring up! We have started a development process here for WFL: https://broadinstitute.github.io/wfl/dev-process/ but it did not specifically mention about commit history. In WFL repo's settings, we forced the commit history to be linear, i.e. only `Squash and merge` button is available and the other 2 are greyed out. Not sure if we could do the same thing for this repo. For now I will squash and merge this PR with it. @hjfbynara

**Comment by rexwangcc** [comment]
Updated the readme in the latest commit, merging the PR now.

**Pull request merged**